### PR TITLE
Source Jira: migration to low-code

### PR DESCRIPTION
**Pull request by roman-yermilov-gl**
## What
Migration from python to low-code for Jira connector.

## How
Most streams have been migrated successfully, except two incremental streams that require incremental synchronization for their parent stream as well. Those streams are:
- IssueComments
- IssueWorklogs

The `PullRequests` stream is also considered to be a Python stream because it requires a lot of customization.

This migration is a breaking change due to the incompatibility between the old and new states. The `LegacyToPerPartitionStateMigration` component cannot be used for the state migration. Additionally, an issue in the `ProjectAvatars` stream has been fixed, where a field was defined as 'project_id,' when in reality, it holds a project key (now it holds project id).

### CODE REVIEW

File: airbyte-integrations/connectors/source-jira/source_jira/components/extractors.py
Comment (artem1205) [inline airbyte-integrations/connectors/source-jira/source_jira/components/extractors.py:28]: Please add unit test to demonstrate why do we really need this component (or remove it):
Existed unit tests can be bypassed using standard DpathExtractor implementation:
image
image


Nit: please use `pytest`'s `ids` to name tests instead of comments.

**Review by artem1205 — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-jira/source_jira/components/extractors.py
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-jira/source_jira/components/extractors.py]: Indeed there is no need in this custom component. This functionality is suported natively. Fixed

**Review by artem1205 — APPROVED** [review]

File: airbyte-integrations/connectors/source-jira/source_jira/components/paginators.py
Comment (girarda) [inline airbyte-integrations/connectors/source-jira/source_jira/components/paginators.py:40]: I'm confused by how this works. Why can't we use a standard limit offset paginator and increment the offset by the size of the the pages?

File: airbyte-integrations/connectors/source-jira/source_jira/streams.py
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-jira/source_jira/streams.py:290]: Done

File: airbyte-integrations/connectors/source-jira/source_jira/components/paginators.py
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-jira/source_jira/components/paginators.py]: The component is going to be removed

File: airbyte-integrations/connectors/source-jira/source_jira/components/paginators.py
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-jira/source_jira/components/paginators.py]: The component is going to be removed

File: airbyte-integrations/connectors/source-jira/source_jira/components/partition_routers.py
Comment (girarda) [inline airbyte-integrations/connectors/source-jira/source_jira/components/partition_routers.py:75]: can you add a comment explaining how this works?

File: airbyte-integrations/connectors/source-jira/source_jira/components/transformations.py
Comment (girarda) [inline airbyte-integrations/connectors/source-jira/source_jira/components/transformations.py:11]: is there value in removing these fields? If anything, [we should go the other way around and make sure all fields are present ](https://airbytehq-team.slack.com/archives/C06FQEBNMEY/p1712778408904579)(doesn't have to be enforced now)

Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-jira/source_jira/components/transformations.py:11]: I can't open the link, but I recall our discussion about this functionality. I mentioned that I would remove it if there is no reason to keep it as it is. However, I found this [PR](https://github.com/airbytehq/airbyte/pull/32927) which provides an explanation of why this feature was added

File: airbyte-integrations/connectors/source-jira/source_jira/components/partition_routers.py
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-jira/source_jira/components/partition_routers.py:88]: Done

File: airbyte-integrations/connectors/source-jira/source_jira/components/paginators.py
Comment (girarda) [reply airbyte-integrations/connectors/source-jira/source_jira/components/paginators.py]: @roman-yermilov-gl do you have an example of an API pagination we can't support? I'd expect CursorPaginator to cover string-based cursor but it's entirely possible we have gaps in our coverage

File: airbyte-integrations/connectors/source-jira/metadata.yaml
Comment (girarda) [inline airbyte-integrations/connectors/source-jira/metadata.yaml:35]:
```suggestion
        upgradeDeadline: "2024-05-31"
```

**Review by girarda — COMMENTED** [review]
Noting I tested the [change here](https://github.com/airbytehq/airbyte/actions/runs/8975268591/job/24649483626) and the result looks good. We should still re-test after the next round of changes but this is moving in the right direction

File: airbyte-integrations/connectors/source-jira/source_jira/config_migrations.py
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-jira/source_jira/config_migrations.py:17]: This code has been unused for a long time before to migration.

File: airbyte-integrations/connectors/source-jira/source_jira/components/paginators.py
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-jira/source_jira/components/paginators.py]: I just was talking about a case when  the next page token is a part of url string:
```
{
    "next": {"nextUrl": "https://example.com?cursor=<the-cursor>"},
    "data": [{ object }]
}
```
In this example I think we can't specify `next.nextUrl` as a cursor field path

File: docs/integrations/sources/jira-migrations.md
Comment (roman-yermilov-gl) [reply docs/integrations/sources/jira-migrations.md:7]: I don't think we can do that because 'per partition' means that we need IDs for each partition. However, for those streams, we only have a common state for parent and child streams, such as:
```
{"updated": "datetime"}
```

File: airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml:989]: I thought if I define an error message it must be returned, but None is returned

File: airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml:156]: Agree, but to avoid blocking the PR I created an issue: https://github.com/airbytehq/airbyte-internal-issues/issues/7604. I will do refatctoring and remove $parameters from the manifest later in current or next sprint

File: airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml:991]: https://github.com/airbytehq/airbyte-internal-issues/issues/7607

File: airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml:1090]: This should be removed

File: airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml:56]: https://github.com/airbytehq/airbyte-internal-issues/issues/7612

File: airbyte-integrations/connectors/source-jira/source_jira/components/partition_routers.py
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-jira/source_jira/components/partition_routers.py:15]: Done

File: airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml:989]: This is an issue https://github.com/airbytehq/airbyte-internal-issues/issues/7613

File: airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml
Comment (girarda) [inline airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml:29]: I'm curious why we're using multiple API versions. Are we in the process of migrating everything to v3?

File: airbyte-integrations/connectors/source-jira/source_jira/components/partition_routers.py
Comment (girarda) [inline airbyte-integrations/connectors/source-jira/source_jira/components/partition_routers.py:55]: why do we need the full record as opposed to the just the parent record id in the slice?

**Review by girarda — APPROVED** [review]
regression tests look good. Approved pending answering the remaining questions and @katmarkham's review of the breaking change notice

File: airbyte-integrations/connectors/source-jira/source_jira/components/partition_routers.py
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-jira/source_jira/components/partition_routers.py:55]: We often require certain data from the parent record to be transferred to the child record. For example, in the HubSpot connector, there are streams where parent streams have composite primary keys, and we need to include all the fields that constitute the primary key into the child record. Here we pass some information to the child record as well, and of course, we don't need all the data. I will optimize this as part of future refactoring.

File: airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml:29]: I'm not entirely sure why, but I suspect it's because the latest version doesn't include an API for retrieving sprint issues, for example.

**Comment by roman-yermilov-gl** [comment]
@girarda I have answered questions and conducted another round of regression tests after all the changes were made, and the results are good.

File: docs/integrations/sources/jira-migrations.md
Comment (katmarkham) [inline docs/integrations/sources/jira-migrations.md:6]: We're continuously striving to enhance the quality and reliability of our connectors at Airbyte. As part of our commitment to delivering exceptional service, we are transitioning source-jira from the Python Connector Development Kit (CDK) to our innovative low-code framework. This is part of a strategic move to streamline many processes across connectors, bolstering maintainability and freeing us to focus more of our efforts on improving the performance and features of our evolving platform and growing catalog. However, due to differences between the Python and low-code CDKs, this migration constitutes a breaking change.

File: docs/integrations/sources/jira-migrations.md
Comment (katmarkham) [inline docs/integrations/sources/jira-migrations.md:8]: To gracefully handle these changes for your existing connections, we highly recommend resetting your data before resuming your data syncs with the new version.

File: docs/integrations/sources/jira.md
Comment (katmarkham) [inline docs/integrations/sources/jira.md:127]: and fix `Project Avatars` stream

File: airbyte-integrations/connectors/source-jira/metadata.yaml
Comment (katmarkham) [inline airbyte-integrations/connectors/source-jira/metadata.yaml:33]: The source Jira connector is being migrated from the Python CDK to our declarative low-code CDK. Due to changes to the incremental stream state, this migration constitutes a breaking change. Additionally, an issue in the `ProjectAvatars` stream has been fixed. After updating, please reset your source before resuming syncs. For more information, see our migration documentation for source Jira.

**Review by katmarkham — CHANGES_REQUESTED** [review]

File: docs/integrations/sources/jira-migrations.md
Comment (roman-yermilov-gl) [reply docs/integrations/sources/jira-migrations.md:8]: Done

File: airbyte-integrations/connectors/source-jira/metadata.yaml
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-jira/metadata.yaml:33]: done

**Review by katmarkham — APPROVED** [review]

**Review by lazebnyi — APPROVED** [review]

**Pull request merged**
